contracts/PreToken.sol

```diff
@@ -61,27 +62,55 @@ contract PreToken is Restricted {
         emit Transfer(0x0, _to, amount);
     }
 
+    /* Restricted function to allow pretoken signers to fix incorrect issuance */
+    function burnFrom(address from, uint amount)
+    public restrict("PreTokenSigner") returns (bool) {
+        require(amount > 0, "burn amount must be > 0"); // this effectively restricts burning from agreement holders only
+        require(agreements[from].balance >= amount, "must not burn more than balance"); // .sub would revert anyways but emit reason
+
+        agreements[from].balance = agreements[from].balance.sub(amount);
+        totalSupply = totalSupply.sub(amount);
+
+        emit Transfer(from, 0x0, amount);
+        return true;
+    }
+
     function balanceOf(address who) public view returns (uint) {
         return agreements[who].balance;
     }
 
     function transfer(address to, uint amount) public returns (bool) { // solhint-disable-line no-simple-event-func-name
+        _transfer(msg.sender, to, amount);
+        return true;
+    }
+
+    /* Restricted function to allow pretoken signers to fix if pretoken owner lost keys */
+    function transferFrom(address from, address to, uint amount)
+    public restrict("PreTokenSigner") returns (bool) {
+        _transfer(from, to, amount);
+        return true;
+    }
+
+    /* private function used by transferFrom & transfer */
+    function _transfer(address from, address to, uint amount) private {
```

Comment:
I would only permit transfer of ownership of the entire agreement and ensure each agreement has a clear ownership (by a single account).

It's not a good idea to try and support ill-defined future use cases such as a secondary market and voting (which seems to be the intent). These can be implemented later, elsewhere, when the requirements are clear.

Reply:
If voting is indeed required, you should specify how you expect it to work. With the current implementation, I could vote multiple times from multiple addresses with a single agreement. Not sure if this is intentional (or in any way better than having a separate voting contract that simply queries if the voter actually holds pretokens).

Reply:
Pushed change to allow only full balance transfer. But note that the original account still can receive further tokens via issueTo because agreement kept at the original account too. I deliberately left it like that because  the account address is included in the signed pretoken agreement.  It's unclear what will be the process changing that address. I.e. we will keep transferring to the original address... This way not too much point to this limitation.

addition to voting: indeed we will decide later if we use this voting mechanism  or do something independently, just reading this contract. Allowing 0 transfers leaves both options open. Can you see any actual issue with it this way?